irohad/simulator/impl/simulator.cpp

```diff
@@ -17,112 +16,72 @@ namespace iroha {
+            validated_proposal_and_errors =
+                validator_->validate(proposal, *storage);
+        ametsuchi_factory_->prepareBlock(std::move(storage));
 
-      auto storage = ametsuchi_factory_->createTemporaryWsv(command_executor_);
-
-      std::shared_ptr<iroha::validation::VerifiedProposalAndErrors>
-          validated_proposal_and_errors =
-              validator_->validate(proposal, *storage);
-      ametsuchi_factory_->prepareBlock(std::move(storage));
-
-      return validated_proposal_and_errors;
-    }
-
-    boost::optional<std::shared_ptr<shared_model::interface::Block>>
-    Simulator::processVerifiedProposal(
-        const std::shared_ptr<iroha::validation::VerifiedProposalAndErrors>
-            &verified_proposal_and_errors,
-        const TopBlockInfo &top_block_info) {
-      const auto &proposal = verified_proposal_and_errors->verified_proposal;
-      if (proposal)
-        log_->info("process verified proposal: {}", *proposal);
-      else
-        log_->info("process verified proposal: no proposal");
-      std::vector<shared_model::crypto::Hash> rejected_hashes;
-      for (const auto &rejected_tx :
-           verified_proposal_and_errors->rejected_transactions) {
-        rejected_hashes.push_back(rejected_tx.tx_hash);
+        return VerifiedProposalCreatorEvent{
+            validated_proposal_and_errors, event.round, event.ledger_state};
+      } else {
+        return VerifiedProposalCreatorEvent{
+            boost::none, event.round, event.ledger_state};
       }
-      std::shared_ptr<shared_model::interface::Block> block =
-          block_factory_->unsafeCreateBlock(top_block_info.height + 1,
-                                            top_block_info.top_hash,
-                                            proposal->createdTime(),
-                                            proposal->transactions(),
-                                            rejected_hashes);
-      crypto_signer_->sign(*block);
-      log_->info("Created block: {}", *block);
-      return block;
     }
 
-    rxcpp::observable<BlockCreatorEvent> Simulator::onBlock() {
-      return block_notifier_.get_observable();
+    BlockCreatorEvent Simulator::processVerifiedProposal(
+        VerifiedProposalCreatorEvent const &event) {
+      if (event.verified_proposal_result) {
+        auto verified_proposal_and_errors = getVerifiedProposalUnsafe(event);
+        auto const &top_block_info = event.ledger_state->top_block_info;
+        auto const &proposal = verified_proposal_and_errors->verified_proposal;
+        if (proposal) {
+          log_->info("process verified proposal: {}", *proposal);
+        } else {
+          log_->info("process verified proposal: no proposal");
+        }
+        std::vector<shared_model::crypto::Hash> rejected_hashes;
+        for (const auto &rejected_tx :
+             verified_proposal_and_errors->rejected_transactions) {
+          rejected_hashes.push_back(rejected_tx.tx_hash);
+        }
```

Comment:
```c++
rejected_hashes.reserve(verified_proposal_and_errors->rejected_transactions.size());
std::transform(verified_proposal_and_errors->rejected_transactions.begin(),
                       verified_proposal_and_errors->rejected_transactions.end(),
                       std::back_inserter(rejected_hashes),
                       [](auto &c) { return c.tx_hash; });

Reply:
or event ```tx_hash``` can be moved.

Reply:
Not sure, because there is a reference to a const event, which contains hashes.